### PR TITLE
feat(exec_command): reject heredoc with missing closing delimiter before spawn

### LINKED ISSUE #1151: feat(exec_command): reject heredoc commands with missing closing delimiter before spawn
## Problem

When a shell heredoc is passed to `exec_command` with a corrupted or missing closing delimiter -- which occurs when JSON escaping turns `\n` into a literal `\n` sequence rather than a newline -- the shell subprocess blocks indefinitely waiting for the delimiter that never arrives. This causes the MCP client (goose) to fire its transport-level timeout (300s or 3600s) instead of receiving a fast, actionable error.

Documented occurrence: [clouatre/dotfiles#677](https://github.com/clouatre/dotfiles/pull/677). The orchestrator wrote a 19 KB JSON payload via a shell heredoc inside `exec_command`. The heredoc delimiter was corrupted by JSON escaping, the shell blocked, and goose fired `-32603: request timeout after PT300S`.

## Root cause

`exec_command` has no pre-spawn validation for heredoc syntax. The command string is passed directly to `sh -c` without checking whether the heredoc body contains the closing delimiter.

Note: `sh -n` (dry-run) does **not** catch this -- POSIX `-n` validates script structure only, not whether heredoc delimiters will be found at runtime. `sh -n -c 'cat <<EOF'` exits 0.

## Proposed fix

In `run_exec_impl` (lib.rs), after `maybe_inject_no_stat` and before `build_exec_command`, scan the post-transform command string for heredoc patterns:

1. Detect `<<[-]?[[:space:]]*` followed by a word-character delimiter (optionally single-quoted, double-quoted, or backslash-escaped per POSIX).
2. Skip `<<` tokens that appear inside a single-quoted region of the command string (e.g. `awk '{x = 1 << 2}'`).
3. Strip quotes from the delimiter to get the bare closing word.
4. For `<<-`, the closing delimiter may be tab-prefixed; strip leading tabs when searching.
5. If the bare delimiter does not appear on its own line (tab-stripped for `<<-`) in the remainder of the command string, reject immediately with `CallToolResult::error` / `INVALID_PARAMS`.

Error hint to include: `"heredoc closing delimiter not found -- likely a quoting or escaping issue; use edit_overwrite to write files instead of shell heredocs"`.

## Properties

- Zero latency (pure string scan, no process spawn)
- No schema changes (no new parameters)
- Non-breaking: only rejects commands that would hang indefinitely anyway
- MCP 2025-11-25 aligned: validation rejections use `CallToolResult::error` with `is_error: true`, same pattern as working_dir and stdin size validation in the same handler

## Acceptance criteria

- `exec_command` with `cat > file.json << 'EOF'\n...\n` (no closing `EOF` line) returns `is_error: true` within milliseconds
- `exec_command` with a valid heredoc (closing delimiter present on its own line) continues to work
- `exec_command` with `awk '{print 1 << 2}'` is not rejected (bitshift inside single-quoted awk program)
- Unit tests covering: valid heredoc, missing delimiter, quoted delimiter variants, `<<-` indented form, single-quoted awk false-positive

### PR DESCRIPTION
**Pull request by clouatre**
## Summary

Add pre-spawn heredoc validation to `exec_command`. Implements `validate_heredocs` in `validation.rs` that scans the command string for unclosed heredoc delimiters before spawning any process. This prevents indefinite hangs when JSON escaping corrupts heredoc delimiters.

The scanner tracks single-quoted and double-quoted regions and skips `<<` tokens inside either, avoiding false positives from `awk` bitshift operators and strings like `echo "use << to redirect"`. For each `<<` or `<<-` token found outside a quoted region, it extracts the bare delimiter word and walks the remaining lines looking for an exact match: no leading whitespace for `<<` (POSIX requires the closer at column 0), only leading tabs stripped for `<<-`, and no trailing whitespace tolerated on either form. This exact-match semantics mirrors what the shell itself enforces.

The fix was validated against a real incident during development: a session attempted to write a file using a shell heredoc passed as a JSON string; the `\n` encoding kept the closing delimiter from ever appearing on its own line, causing the shell to hang and the MCP transport to time out after 300 seconds with a `-32603` error. With this validation in place, that command would have been rejected immediately with a clear `isError: true` message instead.

## Changes

- `crates/aptu-coder/src/validation.rs` -- new `validate_heredocs` function with full quote-region tracking and exact-match closing-delimiter scan
- `crates/aptu-coder/src/lib.rs` -- call `validate_heredocs` in the `exec_command` handler before spawning
- `crates/aptu-coder/tests/exec_command.rs` -- 9 new integration tests: unclosed `<<`, unclosed `<<-`, `<<` inside single-quoted string (no false positive), `<<` inside double-quoted string (no false positive), delimiter on last line with no trailing newline (accepted), unclosed heredoc with no trailing newline (rejected), trailing space on closing line not accepted, leading spaces on closing line not accepted for non-`<<-` heredoc

## Test plan

- [x] Tests pass (34 passed, 0 failed)
- [x] Linter clean (`cargo clippy -- -D warnings`)
- [x] Formatter clean (`cargo fmt --check`)
- [x] All CI checks green

Closes #1151

### CODE REVIEW

File: crates/aptu-coder/src/validation.rs [hunk 1]
```diff
@@ -6,6 +6,170 @@ use rmcp::model::ErrorData;
...
+                    break;
+                }
+
+                if line_end >= len {
+                    // No more lines to check, but if we haven't found the
+                    // delimiter after scanning everything, we'll return Err
+                    // below.
+                    break;
+                }
+
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/validation.rs:158]: CI targets ubuntu-24.04-arm exclusively; CRLF is not a concern here. The scan uses '\n' which is what every tool in this stack produces.

File: crates/aptu-coder/src/validation.rs [hunk 2]
```diff
@@ -6,6 +6,185 @@ use rmcp::model::ErrorData;
...
+                let stripped = if strip_tabs {
+                    line.trim_start_matches('\t')
+                } else {
+                    line
+                };
+
+                if stripped.trim() == delimiter {
+                    found = true;
+                    i = line_end;
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/validation.rs]: Intentional scope: the scan targets the documented hang case (JSON-escaped \\n corrupting the closing delimiter so it never appears on its own line). Nested heredocs and process substitutions are out of scope per #1151. Commands that would bypass this scan would need a closing delimiter present in the string, which means the shell itself would find it and the hang would not occur anyway.

File: crates/aptu-coder/src/validation.rs [hunk 3]
```diff
@@ -6,6 +6,171 @@ use rmcp::model::ErrorData;
...
+        if ch == '\'' {
+            in_single_quote = !in_single_quote;
+            i += 1;
+            continue;
+        }
+
+        // If inside single quotes, skip everything (including `<<` tokens).
+        if in_single_quote {
+            i += 1;
+            continue;
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/validation.rs:50]: Fixed in 64e799d: the main loop now tracks both single-quoted and double-quoted regions. `<<` tokens inside either are skipped, so `echo "use << to redirect"` no longer produces a false positive.

File: crates/aptu-coder/src/lib.rs [hunk 4]
```diff
@@ -3769,6 +3769,13 @@ impl CodeAnalyzer {
             )));
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/lib.rs:3769]: Cannot use `?` here: the enclosing function returns `Result<CallToolResult, ErrorData>`, so `?` would propagate `ErrorData` as the `Err` variant and surface as an MCP-level transport error rather than a `CallToolResult` with `isError: true`. The explicit pattern mirrors the stdin size validation directly above it and is intentional.

File: crates/aptu-coder/src/validation.rs [hunk 5]
```diff
@@ -6,6 +6,179 @@ use rmcp::model::ErrorData;
...
+                let line_end = match command[scan..].find('\n') {
+                    Some(pos) => scan + pos,
+                    None => {
+                        // Last line (no trailing newline)
+                        command[scan..].len() + scan
+                    }
+                };
+
+                let line = &command[scan..line_end];
+
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/validation.rs:141]: Fixed in 42acdc5: replaced the manual index arithmetic with `str::split_inclusive('\n')`. The iterator yields every line (including the final segment with no trailing newline) so advancing past `len` is structurally impossible. Added two integration tests covering the no-trailing-newline edge case: one confirming a valid heredoc with the delimiter on the last line is accepted, and one confirming an unclosed heredoc in the same position is rejected.

File: crates/aptu-coder/src/validation.rs [hunk 6]
```diff
@@ -6,6 +6,168 @@ use rmcp::model::ErrorData;
...
+            // Using split_inclusive('\n') avoids manual index arithmetic and
+            // eliminates any off-by-one risk on the final line: the iterator
+            // yields every line including the terminating '\n' when present, and
+            // the last segment (no trailing newline) is yielded as-is.
+            let mut found = false;
+            let rest = &command[i..];
+            let mut consumed = i;
+
+            for raw_line in rest.split_inclusive('\n') {
+                let line = raw_line.trim_end_matches('\n');
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-coder/src/validation.rs:145]: Fixed in e6a3ccd: replaced `stripped.trim() == delimiter` with an exact equality check. POSIX requires the closing delimiter to appear alone on its line with no leading whitespace (for `<<`) or only leading tabs (for `<<-`), and no trailing whitespace. Verified against bash: `EOF ` (trailing space) and `  EOF` (leading spaces) are not accepted by the shell, and now the scanner rejects them too. Added two integration tests covering both cases.
